### ISSUE TITLE
loading a .trees file with zero-extent spatial bounds?

### ISSUE THREAD

**Issue by petrelharp** [MesserLab/SLiM#176]
Suppose I create and annotate a tree sequence file:
```
import msprime, pyslim

demog_model = msprime.Demography()
demog_model.add_population(initial_size=10)

ots = msprime.sim_ancestry(
    samples=10,
    demography=demog_model,
    random_seed=5,
    sequence_length=100)

ots = pyslim.annotate_defaults(ots, model_type="nonWF", slim_generation=1)
print("Population metadata:")
print(ots.population(0).metadata)
ots.dump("test.trees")
```
which has been annotated to have spatial bounds `0.0`:
```
{'slim_id': 0, 'selfing_fraction': 0.0, 'female_cloning_fraction': 0.0, 'male_cloning_fraction': 0.0, 'sex_ratio': 0.5, 'bounds_x0': 0.0, 'bounds_x1': 0.0, 'bounds_y0': 0.0, 'bounds_y1': 0.0, 'bounds_z0': 0.0, 'bounds_z1': 0.0, 'migration_records': []}
```
and then try to load it into SLiM with:
```
initialize() {
	initializeSLiMModelType("nonWF");
	initializeSLiMOptions(dimensionality="xy");
	initializeTreeSeq();
	
	initializeMutationType("m1", 0.5, "f", 0.0);
	initializeGenomicElementType("g1", m1, 1.0);
	initializeGenomicElement(g1, 0, 99);
	initializeMutationRate(1e-7);
	initializeRecombinationRate(1e-8);
}

1 early() {
   sim.readFromPopulationFile("test.trees");
}

2 late() { catn("done"); }
```
Then, I get the errror:
```
ERROR (SLiMSim::__ConfigureSubpopulationsFromTables): unsorted spatial bounds; this file cannot be read.

Error on script line 19, character 7:

   sim.readFromPopulationFile("test.trees");
```

This only happens if it is a spatial simulation, and it, naturally, just happens because [the bounds are equal](https://github.com/MesserLab/SLiM/blob/master/core/slim_sim.cpp#L8039). I am miffed that my pyslim tests didn't catch this problem. Options are:

1. change pyslim's default annotation to have the upper bounds arbitrary, and greater than the lower bounds
2. change this check so that equality is OK

I vote for the second option, as it should be OK to read in a .trees file and then modify the spatial bounds afterwards.

*Edit:* I first thought another population needed to be added first, but I see that's not the case.

**Comment by bhaller** [comment]
I agree that the error message could be improved – but I didn't anticipate someone trying to do this.  :->  Are you just arguing that the message should be improved, or is there more here?

**Comment by bhaller** [comment]
Ah, I missed your options at the end.  I would opt for #1; I don't see what the meaning would be of a "spatial" simulation in which space has no extent.  I don't think that equality should be OK; if the user wants a bounds-less model, just don't set it up as a continuous-space model in the first place.  Am I missing the point?  Is there a case to be made for allowing this?

**Comment by bhaller** [comment]
Jeez, and then I missed your explanation after the options.  Sorry, not reading carefully, apparently.  Hmm.  Why not just set up the bounds you want in python?  I'm not really comfortable with allowing an invalid model to be read in just so it can be fixed up after the fact on the SLiM side.  At what later point in time would SLiM check that the problem has been fixed?

**Comment by bhaller** [comment]
(I guess the title of this issue should be changed?)

**Comment by petrelharp** [comment]
Three not terribly compelling reasons for allowing zero-size bounds:
1. otherwise I have to pick a default nonzero size for pyslim
2. in a model where someone dynamically resizes the range it might be convenient to not disallow the zero size
3. why not?

**Comment by bhaller** [comment]
> Three not terribly compelling reasons for allowing zero-size bounds:
> 
> 1. otherwise I have to pick a default nonzero size for pyslim
> 2. in a model where someone dynamically resizes the range it might be convenient to not disallow the zero size
> 3. why not?

Hmm.  My replies would be:

1. The default in SLiM is [0,1] in each dimension.  Seems reasonable, and indeed consistent and desirable, for pyslim to use the same default.
2. I don't see why this would really matter; if they're dynamically changing the size anyway, then the initial size doesn't really matter, does it?  They can just change it immediately after load anyway, if they are for some reason unable to set the correct initial bounds on the python side.  That's presumably what they would have done if the bounds had been zero, too, so it makes no difference that I can see.
3. The goal is to help the user avoid bugs.  Zero or inverted bounds seem to be to be meaningless/useless, so they seem like an indicator that the user has forgotten to set up their bounds properly, and it's a helpful service to the user for SLiM to call that to their attention.

Now, yes, the bounds don't really matter for anything; SLiM doesn't actually care what they are.  So we could just say that anything goes.  But by that argument, we might as well allow inverted bounds, too; why not?  I don't buy that.  It is useful to enforce arbitrary limitations sometimes, in order to help keep the user within bounds and help them find bugs in their code.

Back in the day, I used to leave the value of `tag` properties uninitialized, and if the user used a `tag` value before setting it up with a value, well, hey, that's on them.  The behavior was documented; they should have been more careful.  But that led to bugs in many models, and many confused users asking me why their model didn't work.  I changed SLiM to keep track of whether a `tag` value had been initialized or not, and to raise an error on access to a `tag` value before it has been initialized.  Now I get confused users asking me what that error message means – but that's a big improvement, in my book.  :->  There was no need, really, to enforce that limitation; SLiM itself doesn't care at all.  But it's a service to the user.  This strikes me as the same.

If there were an actual use case in which zero bounds was necessary or even particularly useful, then of course it might be a different matter.  But I don't see such a use case being advanced here, and so making it illegal, for debugging purposes, seems helpful to me.

Am I succeeding in convincing you?  :->
